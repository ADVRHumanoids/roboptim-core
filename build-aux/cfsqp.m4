#                                                                -*-Autoconf-*-
# Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
# This file is part of the roboptim.
#
# roboptim is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Additional permission under section 7 of the GNU General Public
# License, version 3 ("GPLv3"):
#
# If you convey this file as part of a work that contains a
# configuration script generated by Autoconf, you may do so under
# terms of your choice.
#
# roboptim is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

# serial 1

m4_pattern_forbid([^CFSQP_])

# CFSQP_LIB([action-if-found], [action-if-not-found])
# --------------------------------------------------------------
# Try to detect the library.
AC_DEFUN([CFSQP_LIB],
#FIXME: replace by cfsqp.
[AC_SEARCH_LIBS([main], [cfsqp], $1, $2)
]) # LIB_CFSQP_LIB

# CFSQP_HEADERS([action-if-found], [action-if-not-found])
# ------------------------------------------------------------------
# Try to detect the library headers.
AC_DEFUN([CFSQP_HEADERS],
[AC_CHECK_HEADERS([cfsqpusr.h], $1, $2)
]) # LIB_CFSQP_HEADERS


# CFSQP_ARG_WITH
# ------------------------------
# Add an optional dependency toward Cfsqp.
AC_DEFUN([CFSQP_ARG_WITH],
[
  AC_ARG_WITH([cfsqp],
  [AS_HELP_STRING([--with-cfsqp],
    [enable cfsqp support])],
  [],
  [with_cfsqp=check])

  AC_SUBST([DISTCHECK_CONFIGURE_FLAGS],
           ["$DISTCHECK_CONFIGURE_FLAGS \
	   '--with-cfsqp=$with_cfsqp'"])

  cfsqp_fail=no
  AS_IF([test "x$with_cfsqp" != xno],
    [
      # Search for Cfsqp library.
      CFSQP_LIB([], [cfsqp_fail=yes])

      # Search for Cfsqp headers.
      CFSQP_HEADERS([], [cfsqp_fail=yes])

      # If both tests are OK, add the library and define the CPP symbol.
      AS_IF([test "x$cfsqp_fail" = xno],
	[AC_SUBST([CFSQP_LIBADD], ["-lcfsqp"])
	 AC_DEFINE([HAVE_CFSQP], [1],
                   [Define if you have cfsqp])])

      # If the support was explicitly required, but detection has failed,
      # this is a fatal error.
      AS_IF([test "x$with_cfsqp" = xyes && \
             test "x$cfsqp_fail" = xyes],
        [AC_MSG_FAILURE(
        [CFSQP test failed (--without-cfsqp to disable)])
        ])
    ])

  # Define a boolean indicating whether or not the package has been found.
  AM_CONDITIONAL([HAVE_CFSQP],
                 [test "x$with_cfsqp" != xno && \
		  test "x$cfsqp_fail" != xyes])
]) # CFSQP_ARG_WITH
