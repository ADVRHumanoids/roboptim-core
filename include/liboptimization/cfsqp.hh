// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Implementation of the CFSQP module.
 */

#ifndef OPTIMIZATION_CFSQP_HH
# define OPTIMIZATION_CFSQP_HH
# include <string>
# include <boost/variant.hpp>
# include <liboptimization/derivable-function.hh>
# include <liboptimization/linear-function.hh>
# include <liboptimization/solver.hh>
# include <liboptimization/util.hh>

namespace optimization
{
  /// CFSQP solver.
  class CFSQPSolver : public Solver<DerivableFunction,
                                    boost::variant<const DerivableFunction*,
                                                   const LinearFunction*> >
  {
  public:
    /// Variant of both Linear and NonLinear functions.
    typedef boost::variant<const DerivableFunction*,
                           const LinearFunction*> constraint_t;

    typedef Solver<DerivableFunction, constraint_t> parent_t;

    /// Constructor.
    explicit CFSQPSolver (const problem_t&, int = 0) throw ();
    /// Destructor.
    virtual ~CFSQPSolver () throw ();
    /// Solve the problem.
    virtual void solve () throw ();
  private:
    /// Initialize bounds.
    void initialize_bounds (double* bl, double* bu) const throw ();

    /// Logging level.
    int iprint_;
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_CFSQP_HH
