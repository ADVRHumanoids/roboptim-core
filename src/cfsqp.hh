// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file src/cfsqp.hh
 *
 * \brief Implementation of the CFSQP module.
 */

#ifndef OPTIMIZATION_CFSQP_HH
# define OPTIMIZATION_CFSQP_HH
# include <string>
# include <solver.hh>

namespace optimization
{
  /// CFSQP solver.
  class CFSQPSolver : public Solver<DerivableFunction, DerivableFunction>
  {
  public:
    typedef Solver<DerivableFunction, DerivableFunction> parent_t;

    /// Constructor.
    explicit CFSQPSolver (const problem_t&, int = 0) throw ();
    /// Destructor.
    virtual ~CFSQPSolver () throw ();
    /// Solve the problem.
    virtual void solve () throw ();
  private:
    /// Initialize bounds.
    void initialize_bounds (double* bl, double* bu) const throw ();

    /// Logging level.
    int iprint_;
  };

} // end of namespace optimization

#endif //! OPTIMIZATION_CFSQP_HH
