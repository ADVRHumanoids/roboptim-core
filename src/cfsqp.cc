// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Implementation of the CFSQP module.
 */
#include <limits>
#include <cfsqpusr.h>

#include <boost/static_assert.hpp>
#include <boost/type_traits/is_base_of.hpp>
#include <boost/variant/apply_visitor.hpp>

#include "liboptimization/cfsqp.hh"
#include "liboptimization/function.hh"
#include "liboptimization/util.hh"

namespace optimization
{
  typedef void (*fct_t) (int, int, double*, double*, void*);
  typedef void (*grad_t) (int, int, double*, double*, fct_t, void*);

  namespace detail
  {
    /// \internal
    /// Evaluate a constraint.
    class EvalConstraintVisitor : public boost::static_visitor<Function::value_type>
    {
    public:
      explicit EvalConstraintVisitor (const CFSQPSolver::problem_t& pb,
                                      const Function::vector_t& x,
                                      int j)
        : pb_ (pb),
          x_ (x),
          j_ (j)
      {}

      template <typename C>
      Function::value_type
      operator () (const C* c)
      {
        BOOST_STATIC_ASSERT((boost::is_base_of<Function, C>::value));

        Function::value_type res = (*c) (x_);

        if (j_ % 2 == 0)
          // g(x) >= b, -g(x) + b <= 0
          res = -res + pb_.bounds ()[j_/2].first;
        else
          // g(x) <= b, g(x) - b <= 0
          res = res - pb_.bounds ()[j_/2].second;
        return res;
      }

    private:
      const CFSQPSolver::problem_t& pb_;
      const Function::vector_t& x_;
      int j_;
    };

    /// \internal
    /// Evaluate a constraint's gradient.
    class EvalGradientConstraintVisitor : public boost::static_visitor<Function::vector_t>
    {
    public:
      explicit EvalGradientConstraintVisitor (const Function::vector_t& x)
        : x_ (x)
      {}

      template <typename C>
      Function::vector_t
      operator () (const C* c)
      {
        BOOST_STATIC_ASSERT((boost::is_base_of<Function, C>::value));

        return c->gradient (x_);
      }

    private:
      const Function::vector_t& x_;
    };


    /// CFSQP objective function.
    void obj (int nparam, int j , double* x, double* fj, void* cd)
    {
      assert (cd);
      CFSQPSolver* solver = static_cast<CFSQPSolver*> (cd);

      Function::vector_t x_ (nparam);
      array_to_vector (x_, x);
      *fj = solver->problem ().function () (x_);
    }

    /// CFSQP constraints function.
    void constr (int nparam, int j,
                 double* x, double* gj, void* cd)
    {
      assert (cd);
      CFSQPSolver* solver = static_cast<CFSQPSolver*> (cd);

      Function::vector_t x_ (nparam);
      array_to_vector (x_, x);

      int j_ = (j < 2) ? 0 : (j - 1)/2;

      EvalConstraintVisitor v (solver->problem (), x_, j);
      *gj = boost::apply_visitor
        (v, solver->problem ().constraints () [j_]);
    }

    /// CFSQP objective function gradient.
    void gradob (int nparam, int j,
                 double* x, double* gradf, fct_t dummy, void* cd)
    {
      assert (cd);
      assert (j == 1);

      CFSQPSolver* solver = static_cast<CFSQPSolver*> (cd);

      Function::vector_t x_ (nparam);
      array_to_vector (x_, x);
      DerivableFunction::gradient_t grad = solver->problem ().function ().gradient (x_);
      vector_to_array (gradf, grad);
    }

    /// CFSQP constraints function gradient.
    void gradcn (int nparam, int j,
                 double* x, double* gradgj, fct_t dummy, void* cd)
    {
      assert (cd);

      CFSQPSolver* solver = static_cast<CFSQPSolver*> (cd);

      Function::vector_t x_ (nparam);
      array_to_vector (x_, x);

      int j_ = (j < 2) ? 0 : (j - 1)/2;

      EvalGradientConstraintVisitor v (x_);

      DerivableFunction::gradient_t grad =
        boost::apply_visitor (v, solver->problem ().constraints ()[j_]);
      vector_to_array (gradgj, grad);
    }

  }

  CFSQPSolver::CFSQPSolver (const problem_t& pb, int iprint) throw ()
    : parent_t (pb),
      mode_ (100),
      iprint_ (iprint),
      miter_ (500),
      bigbnd_ (1e10),
      eps_ (1e-8),
      epseqn_ (1e-8),
      udelta_ (1e-8)
  {
  }

  CFSQPSolver::~CFSQPSolver () throw ()
  {
  }

  void
  CFSQPSolver::initialize_bounds (double* bl, double* bu) const throw ()
  {
    typedef problem_t::bounds_t::const_iterator citer_t;

    Function::size_type i = 0;
    for (citer_t it = problem ().argBounds ().begin ();
         it != problem ().argBounds ().end (); ++it)
      {
        bl[i] = (*it).first, bu[i] = (*it).second;
        ++i;
      }
  }

  void
  CFSQPSolver::solve () throw ()
  {
    int nparam = problem ().function ().n;
    int nf = 1; //FIXME: only one objective function.
    int nfsr = 0;
    int nineqn = 2 * problem ().constraints ().size ();
    int nineq = 2 * problem ().constraints ().size ();
    int neqn = 0;
    int neq = 0;
    int ncsrl = 0;
    int ncsrn = 0;
    int mesh_pts[1];
    int inform = 0;
    double bl[nparam];
    double bu[nparam];
    double x[nparam];
    double f[1];
    double g[2 * problem ().constraints ().size ()];
    double lambda[nparam + 1 + 2 * problem ().constraints ().size ()];
    fct_t obj = detail::obj;
    fct_t constr = detail::constr;
    grad_t gradob = detail::gradob;
    grad_t gradcn = detail::gradcn;

    // Initialize bounds.
    initialize_bounds (bl, bu);

    // Copy starting point.
    if (!!problem ().startingPoint ())
      detail::vector_to_array (x, *problem ().startingPoint ());

    cfsqp (nparam, nf, nfsr, nineqn, nineq, neqn, neq, ncsrl,  ncsrn,
           mesh_pts, mode_,  iprint_, miter_, &inform, bigbnd_, eps_, epseqn_,
           udelta_, bl, bu, x, f, g, lambda,
           obj, constr, gradob, gradcn, this);

    if (inform == 0)
      {
        Function::vector_t x_ (nparam);
        detail::array_to_vector (x_, x);
        result_ = x_;
      }
    else
      result_ = SolverError ("CFSQP has failed.");
  }

  int&
  CFSQPSolver::mode () throw ()
  {
    reset ();
    return mode_;
  }

  const int&
  CFSQPSolver::mode () const throw ()
  {
    return mode_;
  }


  int&
  CFSQPSolver::iprint () throw ()
  {
    reset ();
    return iprint_;
  }

  const int&
  CFSQPSolver::iprint () const throw ()
  {
    return iprint_;
  }


  int&
  CFSQPSolver::miter () throw ()
  {
    reset ();
    return miter_;
  }

  const int&
  CFSQPSolver::miter () const throw ()
  {
    return miter_;
  }

  double&
  CFSQPSolver::bigbnd () throw ()
  {
    reset ();
    return bigbnd_;
  }

  const double&
  CFSQPSolver::bigbnd () const throw ()
  {
    return bigbnd_;
  }

  double&
  CFSQPSolver::eps () throw ()
  {
    reset ();
    return eps_;
  }

  const double&
  CFSQPSolver::eps () const throw ()
  {
    return eps_;
  }

  double&
  CFSQPSolver::epseqn () throw ()
  {
    reset ();
    return epseqn_;
  }

  const double&
  CFSQPSolver::epseqn () const throw ()
  {
    return epseqn_;
  }

  double&
  CFSQPSolver::udelta () throw ()
  {
    reset ();
    return udelta_;
  }

  const double&
  CFSQPSolver::udelta () const throw ()
  {
    return udelta_;
  }

} // end of namespace optimization
