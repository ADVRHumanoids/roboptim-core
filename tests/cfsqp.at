# Copyright (C) 2009 by Thomas Moulard, FIXME.                    -*-Autotest-*-
#
# This file is part of the liboptimization.
#
# liboptimization is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# liboptimization is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

AT_SETUP([Check CFSQPSolver.])

AT_TESTED([cfsqp])
AT_XFAIL_IF([true]) #FIXME: CFSQP is not yet ready.

cp $abs_srcdir/cfsqp.stdout expout
AT_CHECK([cfsqp], [0], [expout], [ignore])

AT_CLEANUP

# End of cfsqp.at
