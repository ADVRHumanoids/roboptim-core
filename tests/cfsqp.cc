// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>
#include <boost/lambda/lambda.hpp>
#include <boost/numeric/ublas/io.hpp>
#include <boost/variant/get.hpp>

#include <liboptimization/cfsqp.hh>

#include "common.hh"
#include "hs071.hh"

int run_test ()
{
  F f;
  G0 g0;
  G1 g1;

  CFSQPSolver::problem_t pb (f);
  pb.getConstraints ().push_back (&g0);
  pb.getConstraints ().push_back (&g1);

  // Set the starting point.
  Function::vector_t start (f.n);
  start[0] = 1., start[1] = 5., start[2] = 5., start[3] = 1.;
  pb.getStartingPoint () = start;

  // Initialize solver
  CFSQPSolver solver (pb, 2);

  // Compute the minimum and retrieve the result.
  CFSQPSolver::result_t res = solver.getMinimum ();

  // Check if the minimization has succeed.
  if (res.which () != CFSQPSolver::SOLVER_VALUE)
    {
      std::cout << "A solution should have been found. Failing..."
                << std::endl;
      return 1;
    }

  // Get the result.
  CFSQPSolver::vector_t& result = boost::get<CFSQPSolver::vector_t> (res);

  // Display the result.
  std::cout << "A solution has been found: " << std::endl;
  std::cout << result << std::endl;
  std::cout << "f(*x) = " << solver.getProblem ().getFunction () (result) << std::endl;
  return 0;
}


GENERATE_TEST ()
