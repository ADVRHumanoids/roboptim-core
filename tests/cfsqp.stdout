

  CFSQP Version 2.5d (Released February 1998) 
          Copyright (c) 1993 --- 1998       
           C.T. Lawrence, J.L. Zhou         
                and A.L. Tits               
             All Rights Reserved            

The given initial point is feasible for inequality
         constraints and linear equality constraints:
                     	   1.00000000000000e+00
 			   5.00000000000000e+00
 			   5.00000000000000e+00
 			   1.00000000000000e+00
 objectives
 			   1.20000000000000e+01
 constraints
 			   0.00000000000000e+00
 			  -1.20000000000000e+01
 
 iteration                                    1
 inform                                       4
 x                   	   1.00000000000000e+00
 			   5.00000000000000e+00
 			   5.00000000000000e+00
 			   1.00000000000000e+00
 objectives
 			   1.20000000000000e+01
 constraints
 			   0.00000000000000e+00
 			  -1.20000000000000e+01
 SNECV               	   1.20000000000000e+01
 step                	   2.22044604925031e-16
 d0norm              	   1.10453610171873e+00
 ktnorm              	   1.10453610171873e+00
 ncallf                                       2
 ncallg                                      57


Error : Step size has been smaller than the computed
machine precision !!


Problem:
  Twice derivable function
  Argument's bounds: (1, 5), (1, 5), (1, 5), (1, 5)
  Argument's scales: 1, 1, 1, 1
  Number of constraints: 2
  Constraint 0
      Bounds: (25, inf)
      Scales: 1
      
  Constraint 1
      Bounds: (40, 40)
      Scales: 1
      
  Starting point: [4](1,5,5,1)
  Infinity value (for all functions): inf
CFSQP specific variables: 
  Nineq: 1
  Nineqn: 1
  Neq: 1
  Neqn: 1
  Mode: 100
  Iprint: 2
  Miter: 500
  Bigbnd: 1e+10
  Eps: 1e-08
  Epseqn: 1e-08
  Udelta: 1e-08
  CFSQP constraints: (0, 1), (1, 1)
A solution should have been found. Failing...
Failed to find a new iterate.
